### ISSUE TITLE
Overwriting date breaks bats tests

### ISSUE THREAD

**Issue by elgohr** [bats-core/bats-core#509]
In https://github.com/elgohr/Publish-Docker-Github-Action I'm overwriting `/usr/bin/date` (https://github.com/elgohr/Publish-Docker-Github-Action/blob/master/Dockerfile#L11) for mocking reasons. Bats 1.2.1 and  previous where totally fine with that.

Since Bats 1.3.0 I'm getting the following errors, before even the setup function is executed:
```
/usr/libexec/bats-core/bats-exec-test: line 127: * 1000 : syntax error: operand expected (error token is "* 1000 ")
/usr/lib/bats-core/tracing.bash: line 61: BATS_STACK_TRACE: bad array subscript
``` 
(see also https://github.com/elgohr/Publish-Docker-Github-Action/issues/134)

When I remove the date-mocking, tests that depend on that mock fail (for the right reason) - but the others work again.

Sadly I can't find anything in the release notes that would point into that direction.
Could you please help me on that?
Any best practise for mocking time in bats? As an alternative I was also looking for a matcher that would take an interval, but didn't find any.

**Comment by jzacsh** [comment]
> I guess for now I'll work around that by providing the two different date formats in the mock.

Can you clarify what the fix is for you? I think the solution was specific to your own test suite: that you found away to _avoid_ getting in bats' way?

_(asking because I just found trying to migrate a 6 year old test suite to bats-core, I'm now hitting this bug in [this mock I have](https://gitlab.com/jzacsh/yabashlib/-/blob/master/spec/mocks/date) [[snapshot](https://gitlab.com/jzacsh/yabashlib/-/blob/ffe040fe02ce01dbb52b2d863c7043fc800d3f88/spec/mocks/date)])_

**Comment by elgohr** [comment]
I'm writing my mocks to a different path (like `/usr/local/mock/`). Then I'm starting the tests without the mocks in the PATH. In `setup()` I'm removing the directory that contains `date` and adding `/usr/local/mock/`

**Comment by elgohr** [comment]
@jzacsh see https://github.com/elgohr/Publish-Docker-Github-Action/commit/99d870b29e035da095b630703e866ae9fc484f3a for example

**Comment by jzacsh** [comment]
> Any best practise for mocking time in bats? 

**tl;dr** I think this^ is probably more important than the `date` bug in particular: just knowing what the (or declaring one) official mocking story is for bats would be helpful. I've re-outlined the bug to demonstrate _one_ variant of `date` being an issue, and left details further down on a concrete example).

## (my understanding of) this mocking bug

I'm realizing links to my own codebase are probably less helpful for someone with just BATs context, so let me explain what I understand about this bug in bats-only terms as bullets here (though my specific case is linked below if someone wants to dig in for something concrete):
- a) [as of this very old copy of bats](https://github.com/sstephenson/bats) (from [2016](https://github.com/sstephenson/bats/tree/03608115df2071fff4eaaff1605768c275e5f81f)) you could mock a basic system utility that your code-under-test relies on (eg if it relies on `date(1)`) directly in your `setup` function before your `@test` cases ran. eg:
- b) create a mock `date` utility: an executable plaintext file with `echo 2021...` inside, somewhere in your repo like `spec/mocks/date`
- c) in your `setup` for a particular `.bats` test file, export `PATH` such that your own `date` will be encountered by bash before the real date, like: `setup () { export PATH=spec/mock:"$PATH"; }`
   - in _reality_ you want to make sure the path is correct (`spec/mock` most certainly wouldn't be) so you'll have some more lines of code than this)
- d) now in your `@test` blocks, when you're done invoking your thing under test, via `run mything.sh`, you now have a reliable output you can check against (eg if a zero-arg call to `mything.sh` is supposed to run `hello world, today is "$(date)"`, then your next `@test` line can be written against your known mock value from step b, eg `hello world, today is 2021...`)

---

<details>
<summary>concrete case: my hack to workaround this issue</summary>

In case it helps anyone else I also figured out a (hacky) workaround: wrap the `run` method. this is hacky because it's brittle and will easily break if anything inside of of the `run` implementation needs access to real date again (or anything else I need to mock in the future): https://gitlab.com/jzacsh/yabashlib/-/blob/bats-issue509-mocks/spec/suite/mocks.bash#L20 ([snapshot](https://gitlab.com/jzacsh/yabashlib/-/blob/8d15eb22ac515a2f982fa4afa4bf79586548bcdc/spec/suite/mocks.bash#L20)). So while this hack is okay for right now, probably most important is to know what someone (with an understanding of bats framework's internals) thinks a sustainable mocking approach should look like.
</details>

**Comment by jzacsh** [comment]
Martin: you mentioned `EPOCHREALTIME`. Did you mean we'd use it inside of bats (instead of using `date` output)? Or did you mean we can use `EPOCHREALTIME` somehow for mocking? If the former, then it sounds like you think replacing major things (like `date`) _should_ work in bats?

**Comment by martin-schulze-vireso** [comment]
Just to chime in on this:

> a) as of this very old copy of bats (from 2016) you could mock a basic system utility that your code-under-test relies on (eg if it relies on date(1)) directly in your setup function before your @test cases ran. eg:

Yes, that old bats version did not use date internally, so you would not have had this problem. However, overriding something even more basic like `ls` would likely lead to problems. In neither version there has been hardening against interactions with the  test code, so mocks can interfere with bats' internals.

For now, you can work around that by using a new shell for mocking instead of mocking inside the bats test itself.

> you mentioned EPOCHREALTIME. Did you mean we'd use it inside of bats (instead of using date output)? Or did you mean we can use EPOCHREALTIME somehow for mocking?

I meant that bats could rely on that variable, thus not requiring date anymore. However, this is only a partial fix for newer bash versions.

**Comment by martin-schulze-vireso** [comment]
Also see this post https://github.com/bats-core/bats-core/issues/230#issuecomment-950175833 However, I am currently considering to at least make an estimate about the costs involved in hardening against this. THis would likely require getting a list of all non bats internal commands that are called by exposed bats parts (everything after test files have been `source`d).

**Comment by jzacsh** [comment]
> Also see this post #230 (comment) 

I agree with what you wrote in that comment - or rather: my perception is that you're saying this isn't sustainable mocking. And I  agree with that very much.

I think better than hardening against this (so as to enable this kind of mocking) would be for us to craft some snippets of how a _test author_ would harden their own code to not leak the mock state quite so much. I took a pass at this when you replied the other day and I just couldn't think of a good[1] way to interject my mock *just* before my own code runs but _after_ bats' `run` func runs.

Has anyone else seen good examples they can point to?

<details>
<summary>[1]: I managed to craft _a_ way to do it, and confirmed it works, but it was terribly hacky. It looks like:</summary>

in my `spec/suite/mocks.bash` file:
```bash
function run_with_mocks() {
  run bash -c 'PATH="$SPEC_MOCKS":"$PATH"; '"$@"
}

function exec_with_mocks() {
  bash -c 'PATH="$SPEC_MOCKS":"$PATH"; '"$@"
}
```

then in my `spec/suite/batsexperiment.bats` file:
```bash
#!/usr/bin/env bats
load mocks

function setup() {
  # normally setup does this work, but now this has to be done in the string that will be executed via `bash -c` since functions and other state of the shell won't be exported unless done so manually
  #  source "$SRCS/my-custom-code-under-test.sh" fooBarTool
}

@test 'isolates the PATH modification for sake of date' {
  MOCK_DATE="$(exec_with_mocks date)"
  run_with_test 'source "$SRCS/my-custom-code-under-test.sh" fooBarTool; myCustomDieFuncUnderTest 99;'
  [[ "$status" -eq 99 ]]
  [[ "${lines[0]}" =~ fooBarTool.*[[:space:]]*$MOCK_DATE$ ]]
}
```

</details>
